cpp/src/arrow/array.cc

```diff
@@ -304,6 +305,55 @@ Status ListArrayFromArrays(const Array& offsets, const Array& values, MemoryPool
   return Status::OK();
 }
 
+std::shared_ptr<Array> SliceArrayWithOffsets(const Array& array, int64_t begin,
```

Comment:
Since this is a local helper function, please mark it `inline` or `static` (can prevent baffling link issues in MSVC)

Reply:
fwiw it's always in an anonymous namespace (thus internal linkage). I'm not familiar with MSVC so I'm still following the suggestion.